clippy_lints/src/significant_drop_tightening.rs

```diff
@@ -58,6 +58,7 @@ impl_lint_pass!(SignificantDropTightening<'_> => [SIGNIFICANT_DROP_TIGHTENING]);
 pub struct SignificantDropTightening<'tcx> {
     /// Auxiliary structure used to avoid having to verify the same type multiple times.
     seen_types: FxHashSet<Ty<'tcx>>,
+    type_cache: FxHashMap<Ty<'tcx>, bool>,
```

Comment:
Why do we need both? Is `type_cache` not enough?

Reply:
We need one for cycle detection (during the recursive call) and one for caching (after the recursive call)